NeoML/src/Dnn/Layers/CrfLayer.cpp

```diff
@@ -123,49 +123,44 @@ bool CCrfCalculationLayer::isFirstStep() const
 
 void CCrfCalculationLayer::RunOnce()
 {
-	int batchWidth = inputBlobs[I_ClassLogProb]->GetBatchWidth();
-	int numberOfClasses = inputBlobs[I_ClassLogProb]->GetObjectSize();
-
 	// The unary estimates of the current elements (the fully-connected layer output)
 	CConstFloatHandle currentProbabilities = inputBlobs[I_ClassLogProb]->GetData();
-	// The vector of partial function value on the previous step (alpha in the forward-backward algorithm, forward pass)
-	// Contains zeros for the first step
-	CConstFloatHandle sequenceProbabilities = inputBlobs[I_ClassSeqLogProb]->GetData();
-	tempSumBlob->Clear();
-	CFloatHandle tempSum = tempSumBlob->GetData();
-	if( !isFirstStep() ) {
+
+	if( isFirstStep() ) {
+		// On the first step all we have to do is initialize the O_ClassSeqLogProb output with current element probabilities
```

Comment:
Why do we skip `outputBlobs[O_BestPrevClass]` here? (if `!IsLearningPerformed()`)

M.b. we should fill it based on `outputBlobs[O_BestPrevClass]` here?

Reply:
Do you mean use of `outputBlobs[O_BestPrevClass] != 0` instead of `!IsLearningPerformed()` to compute `O_BestPrevClass` only when it is connected?
That's a good point. Actually, I have another PR about computing `O_BestPrevClass` and making `CBestSequenceLayer` usable during training, let's discuss it there. I'll publish it right after this one gets merged.